### DISCUSSION TITLE
HeapSize

### DISCUSSION THREAD

**Discussion by fabltd** [bertmelis/espMqttClient#134] [General]
When using this libiary I am seeing a sizable reduction in heapsize.

```
Before MQTT Connection:
free heap size = 99900           min_free_heap_size = 83408 

Connected to MQTT.

 free heap size = 59188           min_free_heap_size = 45432 
```

Any suggestions on how to reduce this.

I am also trying to use the HTTPS OTA but this is crashes due to insufficant free heap. 

Help...

Thanks

**Comment by bertmelis** [comment]
Did you read the issue template?
Hardware? TLS?

**Comment by fabltd** [comment]
My appogies

ESP32-WROOM-32D yes using TLS = espMqttClientSecure mqttClient;

**Comment by bertmelis** [comment]
I guess it's the TLS client. The underlying "client context" is created in heap memory. Not much I can do here.

**Comment by fabltd** [comment]
There are some ways to reduce the TLS client memory any idea where I would set this?

**Comment by bertmelis** [comment]
I don't know, never explored this area.

**Comment by Pablo2048** [comment]
@fabltd if you are on ESP32 there is no easy way AFAIK, because the `MBEDTLS_SSL_MAX_CONTENT_LEN` constant is compile time constant used in all SSL connections and came from `CONFIG_MBEDTLS_SSL_MAX_CONTENT_LEN` from `sdkconfig.h` (that is why i love BearSSL much more over mbedTLS). So the only chance I see is to reconfigure the SDK and rebuild all libraries.

**Comment by proddy** [comment]
BearSSL on an ESP32 has a much lighter footprint than MbedTLS (45kB vs 150kB). Tasmota has a trimmed down arduino/espressif32 core you can use that uses BearSSL. Try replacing the platform in your PIO ini with

platform = https://github.com/tasmota/platform-espressif32.git 

or a specific version like

platform = https://github.com/tasmota/platform-espressif32/releases/download/2023.10.03/platform-espressif32-2023.10.03.zip

**Comment by dzungpv** [comment]
@fabltd If you convert your project to use ESP IDF (It is still Arduino) you can re-config MbedTLS to meet you need, instead using default build, recently I added support this lib to use [ESP IDF](https://github.com/bertmelis/espMqttClient/pull/131), you can see variables to config here: https://docs.espressif.com/projects/esp-idf/en/v4.4.6/esp32/api-reference/kconfig.html#component-config-mbedtls. Update: other solution is use ESP32 with PSRAM.
